File: Week 1/PSet1/credit/credit.c
#include "../../../libcs50-10.1.1/src/cs50.h"
#include <stdio.h>
#include <math.h>

int main(void)
{
    long cardNumber = 0;
    do
    {
        cardNumber = get_long("Number: ");
        printf("%l", get_long("Number: "));
    }
    while (cardNumber < 1);
    // Calculate checksum
    int productSum = 0;
    int nonProductSum = 0;
    long workingNumber = cardNumber;
    int i = 0;
    string result = "INVALID";
    while (workingNumber > 0)
    {
        int currentNumber = workingNumber % 10;
        workingNumber = workingNumber / 10;
        if (i % 2 == 1)
        {
            // Multiply every other digit starting from second-to-last digit with 2
            int product = 2 * currentNumber;
            if (product / 10 == 0)
            {
                productSum += product;
            }
            else
            {
                productSum += product % 10;
                productSum += (product / 10) % 10;
            }
            // Add products together
        }
        else
        {
            // Sum digits that weren't multiplied
            nonProductSum += currentNumber;
        }
        i++;
    }

    // Add sum together
    int numbersSum = productSum + nonProductSum;
    // Check if last digit is 0, if it is, it is valid checksum
    // Check length and starting digits

    // AMEX: 15 digit, 34 || 37
    // MASTERCARD: 16 digit, starts with 51-55
    // VISA: 13 || 16, starts with 4
    // Get card number from input
    if (numbersSum % 10 == 0)
    {
        // Calculate powers for division to find first and second number
        long lengthFirst = pow(10, (i - 1));
        long lengthSecond = pow(10, (i - 2));
        // Calculate first and second number
        int firstNr = cardNumber / lengthFirst;
        int secondNr = (cardNumber / lengthSecond) % 10;
        if (i == 15 && firstNr == 3 && (secondNr == 4 || secondNr == 7))
        {
            // CHECK FOR AMEX
            result = "AMEX";
        }
        else if ((i == 16 || i == 13) && firstNr == 4)
        {
            // CHECK FOR VISA
            result = "VISA";
        }
        else if (i == 16 && firstNr == 5 && (secondNr >= 1 && secondNr <= 5))
        {
            // CHECK FOR MASTERCARD
            result = "MASTERCARD";
        }
    }

    // Print card
    printf("%s\n", result);

}

File: Week 6/Pset6/sentimental-credit/credit.py
import sys
import re
from cs50 import get_string


def is_valid(number):
    # check if card is valid according to Luhn's algorithm
    second_to_last_sum = 0
    not_product_sum = 0
    # traverse provided card number string in reverse
    for i, c in enumerate(reversed(number)):
        # if is odd, starting from reverse. Thus getting second to last
        if i % 2 == 1:
            product = int(c) * 2
            # check if product is smaller than 10
            if product < 10:
                second_to_last_sum += product
            else:
                # if number is greater than 10,
                # add numbers of product separately
                for i in str(product):
                    second_to_last_sum += int(i)
        else:
            # sum together numebers that weren't part of the digit products
            not_product_sum += int(c)
    return (second_to_last_sum + not_product_sum) % 10 == 0


def is_amex(number):
    # check for amex card: starts with 34 or 37 and is 15 characters long
    return re.match('^(34|37)[\d]{13}$', number) and is_valid(number)


def is_mastercard(number):
    # check for mastercard card:
    # starts with 51,52,53,54 or 55 and is 16 characters long
    return re.match('^(51|52|53|54|55)\d{14}$', number) and is_valid(number)


def is_visa(number):
    # check for visa card: starts with 4 and is 13 or 16 characters long
    return re.match('^(4)(\d{12}|\d{15})$', number) and is_valid(number)


def main():
    user_number = get_string('Number: ')
    if is_amex(user_number):
        print("AMEX\n")
    elif is_mastercard(user_number):
        print("MASTERCARD\n")
    elif is_visa(user_number):
        print("VISA\n")
    else:
        print("INVALID\n")


if __name__ == "__main__":
    main()
    sys.exit(0)
